Fix running Fail-Safe cleanup after device is reset during commissioning

adigie opened the pull request:
#### Problem
Due to changes in #20010, when a device is reset during commissioning, Fail-Safe loses information about it. This leads to a situation when after a reset, some partial fabric data is present in the device storage, e.g. following keys:
```
f/1/ac/0/0
f/1/g
f/1/k/0
```
Additionally, if this was first fabric, factory reset is not triggered.

#### Changes
This PR adds `AddNOCStartedMarker` for Fail-Safe which is stored when `AddNOC` command is processed, before any fabric data is stored and deleted when Fail-Safe is disarmed or triggered. Presence of the marker is checked during boot after handling `FabricTable::CommitMarker`, which is still used for keeping consistency of a fabric data during other operations e.g. `UpdateNOC`.

#### Testing
* Added unit tests for `FailSafeContext::AddNOCStartedMarker`.
* Manually verified that all partial fabric data is removed from settings and factory reset is triggered in case this was the only fabric.

adigie committed:
Add PersistentStorageDelegate to FailSafeContext

FailSafeContext needs storage for marker.

Signed-off-by: Adrian Gielniewski <adrian.gielniewski@nordicsemi.no>

adigie committed:
Add AddNOCStartedMarker to FailSafeContext

Signed-off-by: Adrian Gielniewski <adrian.gielniewski@nordicsemi.no>

adigie committed:
Store AddNOCStartedMarker on AddNOC

Signed-off-by: Adrian Gielniewski <adrian.gielniewski@nordicsemi.no>

adigie committed:
Check AddNOCStartedMarker during boot.

Signed-off-by: Adrian Gielniewski <adrian.gielniewski@nordicsemi.no>

adigie committed:
Test Fail-Safe Marker

Test if:
* Marker is stored during AddNOC.
* Marker is not stored during UpdateNOC.
* Marker is cleared During DisarmFailSafe and after timer expiry.

Signed-off-by: Adrian Gielniewski <adrian.gielniewski@nordicsemi.no>

tcarmelveilleux commented on src/app/FailSafeContext.cpp:
```diff
@@ -34,6 +36,57 @@ using namespace chip::DeviceLayer;
+    return CHIP_NO_ERROR;
+}
+
+void FailSafeContext::CheckMarker()
```
The type of marker should be clarified. We used to have a `Commit marker`

tcarmelveilleux replied:
Suggest "FailSafeStartedMarker"

tcarmelveilleux commented on src/app/FailSafeContext.cpp:
```diff
@@ -34,6 +36,57 @@ using namespace chip::DeviceLayer;
+        // Fake arm Fail-Safe and trigger timer expiry.
+        // We handle only the case when new fabric is added. FabricTable CommitMarker
+        // is responsible for guarding the case of updating the existing fabric.
+        SetFailSafeArmed(true);
+        mFabricIndex                 = marker.fabricIndex;
+        mAddNocCommandHasBeenInvoked = true;
+        ForceFailSafeTimerExpiry();
```
Are we sure that when this is called, it's always when FailSafe is not already armed?

adigie replied:
This is called only from `Server::Init`, so it should't be armed at this point. I can add check for that.

tcarmelveilleux commented on src/app/FailSafeContext.cpp:
```diff
@@ -34,6 +36,57 @@ using namespace chip::DeviceLayer;
+{
+    Marker marker;
+
+    CHIP_ERROR err = GetMarker(marker);
```
it is always possible that fail-safe is started for Non-AddNOC reasons. We cannot fake AddNocCommandHasBeenInvoked unless it was actuated, otherwise a marker found while there was no changes yet could lead to removal of a fabric!

adigie replied:
Marker is stored only from `SetAddNocCommandStarted`, so in other scenarios we have previous behavior.

tcarmelveilleux commented on src/credentials/FabricTable.h:
```diff
@@ -413,7 +413,7 @@ class DLL_EXPORT FabricTable
 
     // Returns CHIP_ERROR_NOT_FOUND if there is no fabric for that index.
-    CHIP_ERROR Delete(FabricIndex fabricIndex);
+    CHIP_ERROR Delete(FabricIndex fabricIndex, bool cleanup = false);
```
Please document what the `cleanup` is about

tcarmelveilleux reviewed (changes requested):
The fix here can regress real usecases of fail-safe with no new fabric if reset occurs during fail-safe.

The keys observed were:

```
When a device is reset during commissioning, Fail-Safe loses information about it. This leads to a situation when after a reset, some partial fabric data is present in the device storage, e.g. following keys:

f/1/ac/0/0
f/1/g
f/1/k/0
Additionally, if this was first fabric, factory reset is not triggered and Thread dataset is not cleared.
```

^^^^ This is Group data, Access control data and group key data (IPK). That this implies is that we do not have a way to detect "inconsistent" fabric where we don't have a fabric entry, but there is data still under the fabric that is due to modules that do no properly handle-fail-safe.

#20010 did NOT introduce a regression. #20010 is the CORRECT behavior.

Please engage with me on Slack to resolve this

adigie commented:
> ^^^^ This is Group data, Access control data and group key data (IPK). That this implies is that we do not have a way to detect "inconsistent" fabric where we don't have a fabric entry, but there is data still under the fabric that is due to modules that do no properly handle-fail-safe.

Yes, that's how it works currently if the device is reset somewhere after `AddNOC` and before committing changes in `FabricTable`.

adigie committed:
Don't die when Fail-Safe is armed when checking for AddNOCStartedMarker

This situation should not be possible, however it may lead to bootloop.
Log error and return instead.

Signed-off-by: Adrian Gielniewski <adrian.gielniewski@nordicsemi.no>

adigie committed:
Update logs

Replace `Fail-Safe marker` with `AddNOCStartedMarker`.

Signed-off-by: Adrian Gielniewski <adrian.gielniewski@nordicsemi.no>

adigie commented on src/app/FailSafeContext.cpp:
```diff
@@ -34,6 +36,66 @@ using namespace chip::DeviceLayer;
+    if (err == CHIP_NO_ERROR)
+    {
+        // This should not be possible at this point
+        VerifyOrDie(IsFailSafeArmed() == false);
```
Replaced it with log + return. Also updated some comments to mention `AddNOCStartedMarker` instead of `Fail-Safe marker`
